From 48baf3f6aaec52d61d15dc5cfd8dc6f93aa785ec Mon Sep 17 00:00:00 2001
From: Aprathi K <aprathik@nvidia.com>
Date: Wed, 23 Jul 2025 02:40:52 -0700
Subject: bgpd: Fix default vrf check while configuring md5 password

Issue: bgp dynamic neighbor session configured in default vrf does not
come up, when the bgp listen range is already configured in non-default
vrf.

RC: When setting md5 password for bgp listen range, we iterate
through global list of bgp listeners. We intend to pick the right listener
socket by comparing the vrf and address family.

Now, when md5 password is confiured for bgp neighbor in default vrf,
the existing match condition wrongly picks and sets the md5 on
first listener in the global list having matching address family.

Fix: lister->bgp is NULL for deafult vrf. Including this with
bgp->vrf_id check, should rightly pick the bgp_listener in
default vrf.

Ticket: #4542573
Signed-off-by: Aprathi K <aprathik@nvidia.com>
---
 bgpd/bgp_network.c | 17 ++++++++++++++---
 1 file changed, 14 insertions(+), 3 deletions(-)

diff --git a/bgpd/bgp_network.c b/bgpd/bgp_network.c
index c9b0194247..b173777f69 100644
--- a/bgpd/bgp_network.c
+++ b/bgpd/bgp_network.c
@@ -169,9 +169,20 @@ int bgp_md5_set_prefix(struct bgp *bgp, struct prefix *p, const char *password)
 	/* Set or unset the password on the listen socket(s). */
 	frr_with_privs(&bgpd_privs) {
 		for (ALL_LIST_ELEMENTS_RO(bm->listen_sockets, node, listener))
-			if (listener->su.sa.sa_family == p->family
-			    && ((bgp->vrf_id == VRF_DEFAULT)
-				|| (listener->bgp == bgp))) {
+			/* Match listener socket for the incoming BGP instance if:
+			 * 1. Address family matches (IPv4/IPv6)
+			 * 2. AND either:
+			 *    - incoming BGP instance is in default VRF
+			 *      and listener has no BGP instance(default vrf)
+			 *    - OR listener's BGP instance matches the
+			 *      incoming BGP instance(non-default vrf)
+			 * [Note: listener->bgp is always NULL for default VRF.
+			 *        listener socket could be shared among multiple
+			 *        BGP instances within the same VRF.]
+			 */
+			if (listener->su.sa.sa_family == p->family &&
+			    ((bgp->vrf_id == VRF_DEFAULT && !listener->bgp) ||
+			     (listener->bgp == bgp))) {
 				prefix2sockunion(p, &su);
 				ret = bgp_md5_set_socket(listener->fd, &su,
 							 p->prefixlen,
-- 
2.48.1


From 90dba80480e9ccbd09de3dd0f2b96c2cd3f9be15 Mon Sep 17 00:00:00 2001
From: aprathik <aprathik@nvidia.com>
Date: Thu, 31 Jul 2025 11:36:44 +0000
Subject:  topotest[bgpd]: New bgp auth test case for vrf creation order

 Test addresses issue - bgp dynamic neighbor session with md5 authentication configured in default vrf does not come up,
 when there is bgp listen range is already configured in non-default vrf

    Test steps:
    1. Configure bgp listen range on R1 in blue VRF
    2. Configure R2(blue) and R3(default) to establish bgp session with R1
    3. Verify session between R1 and R2 is established
    4. Configure bgp listen range on R1 in default VRF
    5. Verify bgp sessions (blue, R1, R2) and (default, R1, R3) are established

Signed-off-by: Aprathi K <aprathik@nvidia.com>
---
 .../bgp_auth/R1/frr_dynamic_peers_vrf.conf    |  30 +++
 .../bgp_auth/R2/frr_dynamic_peers_vrf.conf    |  24 +++
 .../bgp_auth/R3/frr_dynamic_peers_vrf.conf    |  20 ++
 tests/topotests/bgp_auth/test_bgp_auth5.py    | 180 ++++++++++++++++++
 4 files changed, 254 insertions(+)
 create mode 100644 tests/topotests/bgp_auth/R1/frr_dynamic_peers_vrf.conf
 create mode 100644 tests/topotests/bgp_auth/R2/frr_dynamic_peers_vrf.conf
 create mode 100644 tests/topotests/bgp_auth/R3/frr_dynamic_peers_vrf.conf
 create mode 100644 tests/topotests/bgp_auth/test_bgp_auth5.py

diff --git a/tests/topotests/bgp_auth/R1/frr_dynamic_peers_vrf.conf b/tests/topotests/bgp_auth/R1/frr_dynamic_peers_vrf.conf
new file mode 100644
index 0000000000..86569106da
--- /dev/null
+++ b/tests/topotests/bgp_auth/R1/frr_dynamic_peers_vrf.conf
@@ -0,0 +1,30 @@
+hostname R1
+
+vrf blue
+ exit-vrf
+!
+
+interface R1-eth0 vrf blue
+ ip address 192.168.1.1/24
+!
+
+interface R1-eth1
+ ip address 192.168.2.1/24
+!
+
+router bgp 65002 vrf blue
+ bgp router-id 10.0.1.1
+ neighbor fabric_pg2 peer-group
+ neighbor fabric_pg2 remote-as internal
+ neighbor fabric_pg2 bfd 3 300 300
+ neighbor fabric_pg2 password blue123
+ neighbor fabric_pg2 advertisement-interval 0
+ neighbor fabric_pg2 timers 1 3
+ neighbor fabric_pg2 timers connect 3
+ neighbor fabric_pg2 capability extended-nexthop
+ # Listen ranges for dynamic neighbors in blue
+ bgp listen limit 500
+ bgp listen range 192.168.1.0/24 peer-group fabric_pg2
+!
+exit
+!
diff --git a/tests/topotests/bgp_auth/R2/frr_dynamic_peers_vrf.conf b/tests/topotests/bgp_auth/R2/frr_dynamic_peers_vrf.conf
new file mode 100644
index 0000000000..727a7f1bb4
--- /dev/null
+++ b/tests/topotests/bgp_auth/R2/frr_dynamic_peers_vrf.conf
@@ -0,0 +1,24 @@
+hostname R2
+
+vrf blue
+ exit-vrf
+!
+
+interface R2-eth0 vrf blue
+ ip address 192.168.1.2/24
+!
+
+router bgp 65002 vrf blue
+ bgp router-id 10.0.1.2
+ neighbor fabric_pg2 peer-group
+ neighbor fabric_pg2 remote-as internal
+ neighbor fabric_pg2 bfd 3 300 300
+ neighbor fabric_pg2 password blue123
+ neighbor fabric_pg2 advertisement-interval 0
+ neighbor fabric_pg2 timers 1 3
+ neighbor fabric_pg2 timers connect 3
+ neighbor fabric_pg2 capability extended-nexthop
+ neighbor 192.168.1.1 peer-group fabric_pg2
+!
+exit
+!
diff --git a/tests/topotests/bgp_auth/R3/frr_dynamic_peers_vrf.conf b/tests/topotests/bgp_auth/R3/frr_dynamic_peers_vrf.conf
new file mode 100644
index 0000000000..8a8ca5da3b
--- /dev/null
+++ b/tests/topotests/bgp_auth/R3/frr_dynamic_peers_vrf.conf
@@ -0,0 +1,20 @@
+hostname R3
+
+interface R3-eth0
+ ip address 192.168.2.2/24
+!
+
+router bgp 65001
+ bgp router-id 10.0.1.3
+ neighbor fabric_pg1 peer-group
+ neighbor fabric_pg1 remote-as internal
+ neighbor fabric_pg1 bfd 3 300 300
+ neighbor fabric_pg1 password default123
+ neighbor fabric_pg1 advertisement-interval 0
+ neighbor fabric_pg1 timers 1 3
+ neighbor fabric_pg1 timers connect 3
+ neighbor fabric_pg1 capability extended-nexthop
+ neighbor 192.168.2.1 peer-group fabric_pg1
+!
+exit
+!
diff --git a/tests/topotests/bgp_auth/test_bgp_auth5.py b/tests/topotests/bgp_auth/test_bgp_auth5.py
new file mode 100644
index 0000000000..85707cea0f
--- /dev/null
+++ b/tests/topotests/bgp_auth/test_bgp_auth5.py
@@ -0,0 +1,180 @@
+#!/usr/bin/env python
+# SPDX-License-Identifier: ISC
+#
+# Copyright (c) 2025 by
+# NVIDIA CORPORATION ("NVIDIA"). All rights reserved.
+#
+#
+
+"""
+test_bgp_auth5.py: Test BGP Md5 Authentication - VRF Prefix Peers Order
+
+Topology:
+    +------------------+     +------------------+
+    |       r1         |     |       r2         |
+    |              blue|-----|blue              |
+    |                  |     |                  |
+    |   default vrf    |     |                  |
+    +------------------+     +------------------+
+           |
+           |
+    +------------------+
+    |   default vrf    |
+    |                  |
+    |       r3         |
+    +------------------+
+
+"""
+# pylint: disable=C0413
+import os
+import platform
+import sys
+
+import pytest
+from lib import common_config, topotest
+from lib.common_config import (
+    step,
+)
+from bgp_auth_common import check_neigh_state
+from lib.topogen import Topogen, get_topogen
+
+pytestmark = [pytest.mark.bgpd]
+
+CWD = os.path.dirname(os.path.realpath(__file__))
+
+
+def build_topo(tgen):
+    tgen.add_router("R1")
+    tgen.add_router("R2")
+    tgen.add_router("R3")
+
+    tgen.add_link(tgen.gears["R1"], tgen.gears["R2"])
+    tgen.add_link(tgen.gears["R1"], tgen.gears["R3"])
+
+
+def setup_module(mod):
+    """Sets up the pytest environment"""
+    # This function initiates the topology build with Topogen...
+    tgen = Topogen(build_topo, mod.__name__)
+    # ... and here it calls Mininet initialization functions.
+    tgen.start_topology()
+
+    r1 = tgen.gears["R1"]
+    r2 = tgen.gears["R2"]
+    r3 = tgen.gears["R3"]
+
+    # blue vrf
+    r1.cmd_raises("ip link add blue type vrf table 1001")
+    r1.cmd_raises("ip link set up dev blue")
+    r2.cmd_raises("ip link add blue type vrf table 1001")
+    r2.cmd_raises("ip link set up dev blue")
+
+    r1.cmd_raises("ip link set R1-eth0 master blue")
+    r2.cmd_raises("ip link set R2-eth0 master blue")
+
+    r1.cmd_raises("ip link set up dev  R1-eth0")
+    r2.cmd_raises("ip link set up dev  R2-eth0")
+
+    r1.cmd_raises("sysctl -w net.ipv4.tcp_l3mdev_accept=1")
+    r2.cmd_raises("sysctl -w net.ipv4.tcp_l3mdev_accept=1")
+    r3.cmd_raises("sysctl -w net.ipv4.tcp_l3mdev_accept=1")
+
+    # This is a sample of configuration loading.
+    router_list = tgen.routers()
+
+    # For all registered routers, load the zebra configuration file
+    for rname, router in router_list.items():
+        router.load_frr_config(
+            os.path.join(CWD, "{}/frr_dynamic_peers_vrf.conf".format(rname))
+        )
+
+    # After copying the configurations, this function loads configured daemons.
+    tgen.start_router()
+
+
+def teardown_module(mod):
+    """Teardown the pytest environment"""
+    tgen = get_topogen()
+    tgen.stop_topology()
+
+
+def verify_peer_states(r1, r2, r3, expected_states):
+    """Verify BGP peer states across all routers and VRFs"""
+    for router, peer_ip, state, vrf in expected_states:
+        check_neigh_state(router, peer_ip, state, vrf)
+
+
+def test_bgp_dynamic_peer_establish_vrf_order(tgen):
+    """
+
+    Test steps:
+    1. Configure bgp listen range on R1 in blue VRF
+    2. Configure R2(blue) and R3(default) to establish bgp session with R1
+    3. Verify session between R1 and R2 is established
+    4. Configure bgp listen range on R1 in default VRF
+    5. Verify session (blue, R1, R2) and (default, R1, R3) is established
+    """
+
+    r1 = tgen.gears["R1"]
+    r2 = tgen.gears["R2"]
+    r3 = tgen.gears["R3"]
+
+    # only supported in kernel > 5.3
+    if topotest.version_cmp(platform.release(), "5.3") < 0:
+        return
+
+    # Step 2: Verify peers are established in VRF blue
+    step("Verify peers are established in VRF blue")
+    bgp_peer_states = [
+        # R2 peer states,
+        (r2, "192.168.1.1", "Established", "blue"),
+        # R1 peer states
+        (r1, "192.168.1.2", "Established", "blue"),
+    ]
+    verify_peer_states(r1, r2, r3, bgp_peer_states)
+
+    # Step 3: Configure peers in default VRF
+    step("Configure peers in default vrf")
+    r1.vtysh_cmd(
+        """
+        configure terminal
+        router bgp 65001
+        bgp router-id 10.0.1.1
+        neighbor fabric_pg1 peer-group
+        neighbor fabric_pg1 remote-as internal
+        neighbor fabric_pg1 bfd 3 300 300
+        neighbor fabric_pg1 password default123
+        neighbor fabric_pg1 advertisement-interval 0
+        neighbor fabric_pg1 timers 1 3
+        neighbor fabric_pg1 timers connect 3
+        neighbor fabric_pg1 capability extended-nexthop
+        bgp listen limit 500
+        bgp listen range 192.168.2.0/24 peer-group fabric_pg1
+    """
+    )
+
+    # Step 4: Verify peers are established in all VRFs
+    step("Verify peers are established in all VRFs")
+    bgp_peer_states = [
+        # R2 peer states,
+        (r2, "192.168.1.1", "Established", "blue"),
+        # R3 peer states
+        (r3, "192.168.2.1", "Established", "default"),
+        # R1 peer states
+        (r1, "192.168.1.2", "Established", "blue"),
+        (r1, "192.168.2.2", "Established", "default"),
+    ]
+    verify_peer_states(r1, r2, r3, bgp_peer_states)
+
+
+def test_memory_leak(tgen):
+    """Run the memory leak test and report results."""
+    if not tgen.is_memleak_enabled():
+        pytest.skip("Memory leak test/report is disabled")
+
+    tgen.report_memory_leaks()
+
+
+if __name__ == "__main__":
+    args = ["-s"] + sys.argv[1:]
+    sys.exit(pytest.main(args))
-- 
2.48.1

